core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -499,7 +500,10 @@ private[log] class Cleaner(val id: Int,
       val timestampType = firstMessageOffset.message.timestampType
       val messageWriter = new MessageWriter(math.min(math.max(MessageSet.messageSetSize(messages) / 2, 1024), 1 << 16))
       messageWriter.write(codec = compressionCodec, timestamp = magicAndTimestamp.timestamp, timestampType = timestampType, magicValue = messageFormatVersion) { outputStream =>
-        val output = new DataOutputStream(CompressionFactory(compressionCodec, outputStream))
+        val compressor = CompressionFactory(compressionCodec, outputStream)
+        if (compressionCodec == LZ4CompressionCodec && magicAndTimestamp.magic == Message.MagicValue_V0)
+          compressor.asInstanceOf[KafkaLZ4BlockOutputStream].useBrokenHC = true // Maintain compatibility with 0.8 / 0.9 clients
```

Comment:
Given that you are mutating a variable of a compression codec in this class after an unsafe cast, things are already very much linked. I think it's much cleaner for `CompressionFactory` to take a message format version and then instantiate the compressor appropriately (that seems like perfectly sensible behaviour for a factory).